nixos/modules/tasks/filesystems/overlayfs.nix

```diff
@@ -0,0 +1,147 @@
+{ config, lib, pkgs, utils, ... }:
+
+let
+  # The scripted initrd contains some magic to add the prefix to the
+  # paths just in time, so we don't add it here.
+  sysrootPrefix = fs:
+    if config.boot.initrd.systemd.enable && (utils.fsNeededForBoot fs) then
+      "/sysroot"
+    else
+      "";
+
+  # Returns a service that creates the required directories before the mount is
+  # created.
+  preMountService = _name: fs:
+    let
+      rwName = "rw-${utils.escapeSystemdPath fs.mountPoint}";
+
+      prefix = sysrootPrefix fs;
+
+      upperdir = prefix + fs.overlay.upperdir;
+      workdir = prefix + fs.overlay.workdir;
+    in
+    lib.mkIf (fs.overlay.upperdir != null)
+      {
+        "${rwName}" = {
+          unitConfig = {
+            DefaultDependencies = false;
+            RequiresMountsFor = "${upperdir} ${workdir}";
+          };
+          serviceConfig = {
+            Type = "oneshot";
+            ExecStart = "${pkgs.coreutils}/bin/mkdir -p -m 0755 ${upperdir} ${workdir}";
+          };
+        };
+      };
+
+  overlayOpts = { config, ... }: {
+
+    options.overlay = {
+
+      lowerdir = lib.mkOption {
+        type = lib.types.nullOr (lib.types.listOf lib.types.path);
```

Comment:
Two suggestions here:
* perhaps use `nonEmptyListOf`?
* `types.path` happily accepts Nix path literals and this is generally a footgun unless you want to copy something to the store.

Reply:
What should I use instead of path then? Just string?

Reply:
After trying to use `types.str`, I think it's actually not very elegant because then you can't just pass a store path as the lowerdir. You'd have to do `lowerdir = [ "${myDerivation}" ]` instead of  `lowerdir = [ myDerivation ]`. Using a store path as a lowerdir, however, is the most common use case IMO.

Reply:
Actually, I now use `either str pathInStore`.